### ISSUE TITLE
How to *continue* pretraining an already pretrained model?

### ISSUE THREAD

**Issue by yuvalkirstain** [bigscience-workshop/Megatron-DeepSpeed#229]
Hey,
 
What should I do to continue pretraining T5/BART/any other arbitrary sequence2sequence pretrained model with this repo (in a way that will be easy to use integrate to Huggingface)? 

If doing so is very tricky, can you please recommend on another way/repository for doing so? I am looking to continue pretraining an already pretrained model on a very large corpus from a different domain.

Any help will be appreciated!

**Comment by thomasw21** [comment]
So we've forked this repo https://github.com/microsoft/Megatron-DeepSpeed 
We currently don't use T5/BART only GPT as it's the only one that have a deepspeed compatible version (search for classes that finish with `Pipe` as suffix)

If you want to train a T5/BART, you can have a look a https://github.com/microsoft/Megatron-DeepSpeed or https://github.com/NVIDIA/Megatron-LM (haven't looked at it, but the code that are linked to theses models in this repo come from those repos)

**Comment by malteos** [comment]
What about GPT models? For example the ones from Huggingface or Megatron (https://catalog.ngc.nvidia.com/orgs/nvidia/models/megatron_lm_345m)? 

Is there any way to load them as checkpoint? The [converter scripts](https://github.com/bigscience-workshop/Megatron-DeepSpeed/tree/main/tools/convert_checkpoint) allow only a conversion from Deepspeed to HF/Meg but not the other way around.

**Comment by thomasw21** [comment]
This was probably generated using `Megatron-LM` which is the original codebase. I believe we still have code that would make this work (though we've introduced some breaking changes along the way, in particular due to positional embeddings). I would suggest using https://github.com/NVIDIA/Megatron-LM instead as this has the better chances of working out of the box.

We don't have a convertion script to convert a Megatron checkpoint to Meg-DS yet.

**Comment by malteos** [comment]
Thanks for the answer. And what about HF models like https://huggingface.co/bigscience/tr5b-1B3-multilingual-alpha-checkpoints/tree/main ?

**Comment by thomasw21** [comment]
HF models can't be loaded in Meg-DS either, it would require writing a conversion script. Is there a reason you would want to use those in Meg-DS instead of `transformers`?

**Comment by malteos** [comment]
I'm already training models with Meg-DS from scratch but now I want to reuse existing pretrained checkpoints to make the training more efficient. For training I want to stick to Meg-DS for scalability reasons but unfortunately the pretrained checkpoints are in HF.  

Apart from the weight name mapping, what other differences are there?

**Comment by thomasw21** [comment]
I see. So I haven't played around with it but out of the top of my head:
 - You need to generate the correct structure. Meg-DS checkpoints have multiple files, whereas HF pretrained models are stored inside a single bin:
     - `layer_{layer_id}-model_{tp_rank}-model_states.pt` to store single layers individually. `layer_id` is it's index in the sequence of layers. There's probably a reshaping loading mechanism allowing you to just target TP=1.
     - `mp_rank_{tp_rank + pp_rank * TP}_model_states.pt` to store a bunch of meta datas. 
 - You can probably ignore optim states through an argument. https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/4c13c617bdfb3bd419afd620bae87c74ae5aa79d/megatron/arguments.py#L655-L656

**Comment by malteos** [comment]
I've tried replacing the weights with `load_state_dict` and mapping the corresponding layer names. See [here](https://github.com/malteos/bigscience-Megatron-DeepSpeed-oxw/blob/oxw/pretrain_gpt.py#L86-L96) and [here](https://github.com/malteos/bigscience-Megatron-DeepSpeed-oxw/blob/oxw/megatron/model/from_pretrained_hf.py#L68). 

After training with Bigscience-DS-Meg for a single iteration and tiny LR, the saved checkpoint is almost identical to the HF checkpoint. So it seems to work on the weight-level. However the LM loss does not match. 

What could cause this? Is due to the different implementation of activation functions and layer norms?

**Comment by malteos** [comment]
In case somebody has the same problem: 

Even though the checkpoint weights seemed identical - they were not. The self attention weight/bias needed to be reordered (basically reverting the `fix_query_key_value_ordering` from `convert_megatron_gpt2_checkpoint.py`. see [here](https://github.com/malteos/bigscience-Megatron-DeepSpeed-oxw/blob/oxw/megatron/model/from_pretrained_hf.py#L95-L109)). Having fixed this, continue pretraining from the BigScience 1B HF checkpoint works but only with TP/PP=1. Other topologies would require additional changes.

**Comment by thomasw21** [comment]
Sorry missed your last message:

> After training with Bigscience-DS-Meg for a single iteration and tiny LR, the saved checkpoint is almost identical to the HF checkpoint. So it seems to work on the weight-level. However the LM loss does not match.

Interesting, can you confirm that the EXACT same weights don't generate the same loss (in which case we might have a bug somewhere). Activation can definitely play a role, see https://arxiv.org/abs/2002.05202 for example. Layer norm has been fixed in pytorch 1.11 (so I recommend updating to it). Essentially it allows to accumulate in fp32 instead fp16, see https://github.com/pytorch/pytorch/pull/66920

If you have a script to load a HF transformer model inside Meg-DS feel free to open a PR. I don't think it's something we planned on doing, but I think it's valuable enough to be added to this repo.

**Comment by malteos** [comment]
> Interesting, can you confirm that the EXACT same weights don't generate the same loss (in which case we might have a bug somewhere). Activation can definitely play a role, see https://arxiv.org/abs/2002.05202 for example. Layer norm has been fixed in pytorch 1.11 (so I recommend updating to it). Essentially it allows to accumulate in fp32 instead fp16, see [pytorch/pytorch#66920](https://github.com/pytorch/pytorch/pull/66920)
> 

The weight weren't exactly the same (see my last comment). The wrong order of the attention QKV weights caused the problem. I didn't notice this at first because the shapes and the first/last weights matched.

As soon as I have a through out tested version I'll do a PR ;-)
